### PR TITLE
Allow Tabbing between TextBoxes

### PR DESCRIPTION
**Pull request by 2yangk23**
Only added for TextBox, since I was looking at other controls and they don't seem to get focused like TextBoxes do.  Let me know if you think it should be made more generic for all controls.

Tested it with the login form, just need to add:
```
username.TabNext = password;
username.TabPrev = password;
password.TabNext = username;
password.TabPrev = username;
```

### CODE REVIEW

**Comment by SirCmpwn** [comment]
Would it be better to have InputManager walk up the tree and look for other inputable fields?

**Comment by 2yangk23** [comment]
I only looked at the code for like 30min so I don't have full understanding, but from what I can tell, InputManager is some top level input manager.  What I would ideally need is some way to access the parent of the focused form, so you only look for inputable fields on the same level.  Otherwise you would be able to tab to all kinds of places rather than cycling through the fields you are interested in.

From a child control, I don't see any way of accessing the parent.  Maybe there could be a reference kept?

**Comment by peppy** [comment]
One issue I can see here is it only traverses ont parent level. In options for instance, textboxes are one extra level deep, which would break this.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 1]
```diff
@@ -256,6 +260,33 @@ private int getCharacterClosestTo(Vector2 pos)
 
         Cached<Vector2> cursorAndLayout = new Cached<Vector2>();
 
+        private Drawable nextTabStop(IReadOnlyList<Drawable> children, int current, bool reverse)
+        {
+            if (reverse)
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: wouldn't it be easier to just do 
```var filtered = children.FindAll(t => t is TextBox);
var current = c.IndexOf(this);
return filtered[(current + 1) % filtered.Count]```

**Comment by peppy** [comment]
To get around the limitation I mentioned, I recommend you make an interface

```
ITabbable
{
    Container TabbableContainer { get; set; }
} 
```

Then you can make TextBox inherit this and change your search methods to look for ITabbables (not just limited to TextBoxes). It will also avoid using `Parent`, which is something we don't want to do whenever possible.

File: osu.Framework/Graphics/UserInterface/TextBox.cs (hunk 1 above)
Comment (2yangk23) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: Yea you're right, I was just trying to not create any new lists, but that's over optimizing and hard to read. I'll change it.

**Comment by 2yangk23** [comment]
> One issue I can see here is it only traverses on parent level. In options for instance, textboxes are one extra level deep, which would break this.

Do you think it is appropriate for the search to handle this?  I feel that allowing TAB to move between multiple levels will cause unwanted behavior in most cases.  Generally, I feel you shouldn't be able to tab out of your current level since that may cause some inconsistent behavior.

Example: You are tabbing in the login form, and it focuses an options control.

**Comment by peppy** [comment]
Did you see my next comment `ITabbable`? In this usage, you would set the TabbableContainer property to the "target" container which you want to search for tabbables.

You'll need to search recursively, though.

**Comment by 2yangk23** [comment]
Ah, I get it, I misunderstood the usage of the target at first.  (Thought it was some reference to the same control)

**Comment by peppy** [comment]
might not be the best name. just make sure to comment it well :)

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 2]
```diff
@@ -466,6 +466,12 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
             switch (args.Key)
             {
                 case Key.Tab:
+                    var next = NextTabStop(state.Keyboard.ShiftPressed);
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: tab handling logic should go inside TabbableContainer.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace osu.Framework.Graphics.Containers
+{
+    public class TabbableContainer : TabbableContainer<Drawable> { }
+
+    public class TabbableContainer<T> : Container<T> 
+        where T : Drawable
+    {
+        /// <summary>
+        /// Base container for tabbing between ITabbable. Allows for tabbing between
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: You aren't even using an interface, so don't mention it here.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 4]
```diff
@@ -466,6 +466,12 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
             switch (args.Key)
             {
                 case Key.Tab:
+                    var next = NextTabStop(state.Keyboard.ShiftPressed);
+                    if (next != null) {
+                        TriggerFocusLost();
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: i believe this is unnecessary.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs (hunk 3 above)
Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: oops, that was leftover. Removed.

File: osu.Framework/Graphics/UserInterface/TextBox.cs (hunk 4 above)
Comment (2yangk23) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: yep, removed.

**Comment by peppy** [comment]
Looking pretty usable to me 👍 .

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+{
+    public class TabbableContainer : TabbableContainer<Drawable> { }
+
+    public class TabbableContainer<T> : Container<T> 
+        where T : Drawable
+    {
+        /// <summary>
+        /// Allows for tabbing between multiple levels within the BaseContainer.
+        /// </summary>
+        public Container<Drawable> BaseContainer { private get; set; }
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: might call the TabbableContentContainer

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+        {
+            if (args.Key != Key.Tab)
+                return false;
+            var next = nextTabStop(BaseContainer, state.Keyboard.ShiftPressed ? -1 : 1);
+            next?.TriggerFocus();
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: can be combined into a single line without variable assignment.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            return true;
+        }
+
+        private Drawable nextTabStop(Container<Drawable> target, int step, bool intialized = false)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: initialized isn't a great variable name here.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            if (intialized && target is TabbableContainer)
+                return target;
+
+            LifetimeList<Drawable> children = target?.Children as LifetimeList<Drawable>;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: is this case really necessary?

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            if (children == null)
+                return null;
+
+            var filtered = children.FindAll(t => t is Container<Drawable>)
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: more FindAll and use OfType

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                .Cast<Container<Drawable>>()
+                .ToList();
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: i don't think you need ToList either.

Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: Then it becomes an IEnumerable and I cannot access by index.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs (hunk 8 above)
Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: It's the base case.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs (hunk 6 above)
Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: I think it's more readable this way

Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: it's not though

File: osu.Framework/Graphics/Containers/TabbableContainer.cs (hunk 8 above)
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: sorry, i meant cast.

Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: Needed it for FindAll at first, but using OfType it's no longer necessary.

**Comment by peppy** [comment]
is there another commit you have pending for this? i'll take another look once you've fixed the parts you want to fix.

**Comment by 2yangk23** [comment]
I was at work, but I'll make the changes now.

File: osu.Framework/Graphics/Containers/TabbableContainer.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+            bool started = false;
+            while (stack.Count > 0)
+            {
+                var container = stack.Pop();
+                if (container == null)
+                    continue;
+                
+                if (!started)
+                    started = ReferenceEquals(container, this);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: shouldn't need ReferenceEquals here, just `container == this`

Comment (2yangk23) [inline osu.Framework/Graphics/Containers/TabbableContainer.cs]: I have to, it gives this:

> Operator '==' cannot be applied to operands of type 'Container<Drawable>' and 'TabbableContainer<T>'	osu.Framework

Alternatively, I considered

`container == this as TabbableContainer`

but I prefer this solution.

**Comment by peppy** [comment]
Can you update `TestCaseTextBox` to allow testing of tab functionality?

**Comment by 2yangk23** [comment]
rip I found a bug while adding some more tests just so you know.

**Comment by peppy** [comment]
looks good 👍

**Pull request merged**